File: java/src/test/java/neetcode/slidingwindow/p0076_minimumwindowsubstring/SolutionTest.java
package neetcode.slidingwindow.p0076_minimumwindowsubstring;

import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;

import static org.junit.jupiter.api.Assertions.assertEquals;

class SolutionTest {

    private Solution sol;

    @BeforeEach
    void setUp() {
        sol = new Solution();
    }

    @Test
    void test_neetcode_example() {
        assertEquals("YXAZ", sol.minWindow("OUZODYXAZV", "XYZ"));
    }

    @Test
    void test_classic_leetcode_example() {
        assertEquals("BANC", sol.minWindow("ADOBECODEBANC", "ABC"));
    }

    @Test
    void test_single_char_match() {
        assertEquals("a", sol.minWindow("a", "a"));
    }

    @Test
    void test_no_valid_window() {
        // t has a character not in s
        assertEquals("", sol.minWindow("a", "aa"));
    }

    @Test
    void test_t_longer_than_s() {
        assertEquals("", sol.minWindow("ab", "abc"));
    }

    @Test
    void test_whole_string_is_minimum() {
        // Every character of t appears exactly once in s, spread across it
        assertEquals("AECBAD", sol.minWindow("AECBAD", "ABCDE"));
    }

    @Test
    void test_duplicates_in_t() {
        // t = "aa" requires at least two 'a's in the window
        assertEquals("aa", sol.minWindow("aa", "aa"));
        assertEquals("", sol.minWindow("a", "aa"));
        assertEquals("baa", sol.minWindow("baabc", "aa"));
    }

    @Test
    void test_multiple_valid_windows() {
        // "ab" and "ba" are both valid but "ab" (len 2) beats "abc" (len 3)
        // s = "abc", t = "ac" → minimum is "abc" (length 3) wait:
        // actually "a" is at 0, "c" is at 2, so minimum window is "abc" length 3
        assertEquals("abc", sol.minWindow("abc", "ac"));
    }

    @Test
    void test_empty_s() {
        assertEquals("", sol.minWindow("", "a"));
    }

    @Test
    void test_empty_t() {
        assertEquals("", sol.minWindow("abc", ""));
    }
}

File: python/src/neetcode/slidingwindow/p0076_minimumwindowsubstring/solution_test.py
from solution import Solution


class TestSolution:
    def setup_method(self):
        self.sol = Solution()

    def test_neetcode_example(self):
        assert self.sol.minWindow("OUZODYXAZV", "XYZ") == "YXAZ"

    def test_leetcode_classic(self):
        assert self.sol.minWindow("ADOBECODEBANC", "ABC") == "BANC"

    def test_single_char_match(self):
        assert self.sol.minWindow("a", "a") == "a"

    def test_no_valid_window(self):
        assert self.sol.minWindow("a", "aa") == ""

    def test_t_longer_than_s(self):
        assert self.sol.minWindow("ab", "abc") == ""

    def test_empty_t(self):
        assert self.sol.minWindow("abc", "") == ""

    def test_empty_s(self):
        assert self.sol.minWindow("", "a") == ""

    def test_duplicates_in_t(self):
        assert self.sol.minWindow("aa", "aa") == "aa"
        assert self.sol.minWindow("a", "aa") == ""

    def test_whole_string_is_minimum(self):
        assert self.sol.minWindow("AECBAD", "ABCDE") == "AECBAD"
